analyses/npl/10_historical_event_timeline.md

```diff
@@ -0,0 +1,307 @@
+                     right_index=True, 
+                     suffixes=['_wl', '_rd'])
+            #.dropna()
+            .rename(columns={f"{station}_wl": "water_level",
+                    f"{station}_rd": "river_discharge"})
+           )
+    # Fill in the gaps so that the group finding works - not sure this is needed
+    data = data.reindex(pd.date_range(data.index.min(), data.index.max()))
+    # Add in the forecast data
+    for leadtime in LEADTIMES:
+        for percentile in FORECAST_PERCENTILE_LIST:
+            forecast = (ds_glofas_forecast_summary[station + VERSION_LOC]
+                    .sel(leadtime=leadtime, percentile=percentile)
+                    .to_dataframe()
+                    .drop(columns=['surface', 'leadtime', 'percentile'])
+                    .rename(columns={f"{station+VERSION_LOC}": station}))
+            data = (pd.merge(data, forecast,
+                        how='left',
+                        left_index=True,
+                        right_index=True,
+                        )
+               .rename(columns={station: f"forecast_lt{leadtime}_p{percentile}"}))
+    # Keep only the forecast date range
+    data = data.dropna(subset=["forecast_lt1_p50"])
+    # Get the water level events
+    level_type = "danger"
+    level_val = df_station_info.at[station, f'{level_type}_level']
+    events = utils.get_groups_above_threshold(data['water_level'], level_val, min_duration=DURATION)
+    event_start_indices = [event[0] for event in events]
+    data[f"event_{level_type}"] = False
+    data[f"event_{level_type}"].iloc[event_start_indices] = True
+    # Get river discharge events
+    for rp in RP_LIST:
+        rp_val = df_return_period.loc[rp, station]
+        events = utils.get_groups_above_threshold(data['river_discharge'], rp_val, min_duration=DURATION)
+        event_start_indices = [event[0] for event in events]
+        data[f"event_rp{rp}"] = False
+        data[f"event_rp{rp}"].iloc[event_start_indices] = True
+    # Go through each date and add activations
+    for rp in RP_LIST:
+        rp_val = df_return_period.loc[rp, station]
+        for percentile in FORECAST_PERCENTILE_LIST:
+            for event_type, leadtimes in LEADTIMES_BY_TRIGGER.items():
+                data[f'event_{event_type}_rp{rp}_p{percentile}'] = data.apply(lambda row: 
+                                          (row[[f'forecast_lt{leadtime}_p{percentile}' for leadtime in leadtimes]] >= rp_val).any(),
+                                          axis=1)
+    df_station_dict[station] = data
+data
+```
+
+### Get TP, FP, and FN
+
+```python
+# Find TP, FP and FN
+
+def get_consecutive_groups(x: pd.Series, n=2):
+    # Find indices whre False / True change. Skip the first entry of the array because it always changes there
+    l = np.where(x.shift() != x)[0][1:]
+    # Convert array to pairs
+    return [l[i:i+n] for i in range(0, len(l), n)]
+
+
+def get_station_stats(df_station_dict, event_var, rp=MAIN_RP):
```

Comment:
I think this would definitely be good to generalize and add to the general glofas ```utils.py``` (probably improving on what I've done in #135).

Reply:
Agreed 110%, new PR coming up ASAP!